resource-recorder/src/lib.rs

```diff
@@ -134,7 +134,26 @@ where
     ) -> Result<Response<ResultResponse>, Status> {
         request.verify(Scope::ResourcesWrite)?;
 
+        let can_provision_rds = request.verify_rds_access();
+
         let request = request.into_inner();
+
+        if let Err(error) = can_provision_rds {
+            if request.resources.iter().any(|resource| {
+                let Ok(r#type) = shuttle_common::database::Type::from_str(resource.r#type.as_str())
+                else {
+                    return false;
+                };
+
+                matches!(r#type, shuttle_common::database::Type::AwsRds(_))
+            }) {
+                return Ok(Response::new(ResultResponse {
+                    success: false,
+                    message: error.message().to_string(),
+                }));
+            }
+        };
```

Comment:
Yes, the provisioning failure will cause an error leading to a failed deployment, and it won't get to the recording step.

Having the resource request take a vec makes sense for the case when we want to send cached resources from old deployers to r-r without having to send multiple requests, but that will become more and more rare. In the common case where it is called in the runtime resource tracker, it is called once for each resource.

This does lead me to consider the case where someone starts an older deployer and it tries to send the resources to r-r, if they have an RDS and they are not on the pro tier, the deployment will fail. But the database will still be there. I'm not sure this is desired behavior, at least not at this step, since they will then be locked out (if they don't have the connection string handy) before they can migrate away. I think we should consider leaving the r-r check out of this first iteration. :thinking:

Reply:
Reverted the r-r changes in [e6ba150](https://github.com/shuttle-hq/shuttle/pull/1398/commits/e6ba150c39dbf33511ac7ee7fbfc160255207258), and verified that deployment still fails on provisioning with permission denied error.